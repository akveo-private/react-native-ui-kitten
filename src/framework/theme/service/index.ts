export {
  Provider,
  Consumer,
} from './reactContext.service';
export { forwardProps } from './reactPropsForward.service';
