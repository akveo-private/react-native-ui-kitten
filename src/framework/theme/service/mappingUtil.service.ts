import {
  ThemeMappingType,
  ComponentMappingType,
  AppearanceType,
  VariantGroupType,
  MappingType,
  StateType,
} from '../component';

export const APPEARANCE_DEFAULT = 'default';

/**
 * @param component: string - component name
 * @param mapping: ThemeMappingType - theme mapping configuration object
 *
 * @return ComponentMappingType if presents in mapping, undefined otherwise
 */
export function getComponentMapping(mapping: ThemeMappingType,
                                    component: string): ComponentMappingType | undefined {

  return mapping[component];
}

export function getAppearance(mapping: ComponentMappingType,
                              appearance: string): AppearanceType | undefined {

  return mapping.appearance[appearance];
}

export function getAppearanceMapping(mapping: ComponentMappingType,
                                     appearance: string): MappingType | undefined {

  const appearanceConfig = getAppearance(mapping, appearance);

  return appearanceConfig && appearanceConfig.mapping;
}

export function getAppearanceMappingSafe(mapping: ComponentMappingType,
                                         appearance: string,
                                         fallback: MappingType): MappingType {

  return getAppearanceMapping(mapping, appearance) || fallback;
}

export function getAppearanceVariants(mapping: ComponentMappingType,
                                      appearance: string): VariantGroupType | undefined {

  const appearanceConfig = getAppearance(mapping, appearance);

  return appearanceConfig && appearanceConfig.variant;
}

export function getVariantMapping(mapping: ComponentMappingType,
                                  appearance: string,
                                  variant: string): MappingType | undefined {

  const variantGroupConfig = getAppearanceVariants(mapping, appearance);
  const variantGroupName = variantGroupConfig && Object.keys(variantGroupConfig).find(group => {
    return variantGroupConfig[group][variant] !== undefined;
  });
  const variantConfig = variantGroupName && variantGroupConfig[variantGroupName][variant];

  return variantConfig && variantConfig.mapping;
}

export function getVariantMappingSafe(mapping: ComponentMappingType,
                                      appearance: string,
                                      variant: string,
                                      fallback: MappingType): MappingType {

  return getVariantMapping(mapping, appearance, variant) || fallback;
}

export function getMappingState(mapping: MappingType,
                                state: string): StateType | undefined {

  return mapping.state && mapping.state[state];
}
