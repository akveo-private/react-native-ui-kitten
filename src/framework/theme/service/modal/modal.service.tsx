import React from 'react';
import { ModalPanel } from '../../component';

class ModalServiceType {

  component: ModalPanel | null = null;

  setComponent(component: ModalPanel | null): void {
    this.component = component;
  }

  showDialog(dialogComponent: React.ReactElement<any>, closeOnBackDrop: boolean = false): void {
    if (this.component) {
      this.component.showDialog(dialogComponent, closeOnBackDrop);
    }
  }
}

export const ModalService = new ModalServiceType();
