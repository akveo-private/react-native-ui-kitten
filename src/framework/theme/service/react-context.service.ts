import React from 'react';
import { ThemeShape } from '../component';

const defaultThemeValue: ThemeShape = {};

const {
  Provider,
  Consumer,
} = React.createContext(defaultThemeValue);

export {
  Provider,
  Consumer,
};
