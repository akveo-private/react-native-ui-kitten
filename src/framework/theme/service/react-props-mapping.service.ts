import React, { ComponentType } from 'react';

const REACT_METHODS = [
  'autobind',
  'childContextTypes',
  'componentDidMount',
  'componentDidUpdate',
  'componentWillMount',
  'componentWillReceiveProps',
  'componentWillUnmount',
  'componentWillUpdate',
  'contextTypes',
  'displayName',
  'getChildContext',
  'getDefaultProps',
  'getDOMNode',
  'getInitialState',
  'mixins',
  'propTypes',
  'render',
  'replaceProps',
  'setProps',
  'shouldComponentUpdate',
  'statics',
  'updateComponent',
];

export function forwardProps<P>(Source: ComponentType<P>, Target: ComponentType<P>): ComponentType<P> {
  function filterProps(prop) {
    // React specific methods and properties or properties from React's prototype
    const isReactProp = REACT_METHODS.includes(prop) || prop in React.Component.prototype;
    // Properties from enhanced component's prototype
    const isTargetProp = prop in Target.prototype;
    // Private methods
    const isPrivateProp = prop.startsWith('_');

    return !(isReactProp || isTargetProp || isPrivateProp);
  }

  function mapProps(prop) {
    if (typeof Source.prototype[prop] === 'function') {
      // Make sure the function is called with correct context
      Target.prototype[prop] = function (...args) {
        return Source.prototype[prop].apply(this.getConsumingComponent(), args);
      };
    } else {
      // Copy properties as getters and setters
      Object.defineProperty(Target.prototype, prop, {
        get() {
          return this.getConsumingComponent()[prop];
        },
        set(value) {
          this.getConsumingComponent()[prop] = value;
        },
      });
    }
  }

  Object.getOwnPropertyNames(Source.prototype).filter(filterProps).forEach(mapProps);

  return Target;
}
