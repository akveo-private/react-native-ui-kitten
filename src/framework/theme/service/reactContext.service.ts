import React from 'react';
import { ThemeType } from '../component';

const defaultThemeValue: ThemeType = {};

const {
  Provider,
  Consumer,
} = React.createContext(defaultThemeValue);

export {
  Provider,
  Consumer,
};
