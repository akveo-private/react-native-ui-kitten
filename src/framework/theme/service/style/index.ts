export * from './style.service';
export * from './styleConsumer.service';
