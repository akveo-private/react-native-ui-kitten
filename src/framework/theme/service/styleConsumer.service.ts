import { ComponentMappingType } from '../component';
import { Props } from '../component/style/styleConsumer.component';
import {
  APPEARANCE_DEFAULT,
  getAppearanceVariants,
} from './mappingUtil.service';

export class StyleConsumerService {

  public getVariantPropKeys<P extends Props>(mapping: ComponentMappingType, props: P): string[] {
    const variants = getAppearanceVariants(mapping, APPEARANCE_DEFAULT);
    if (variants === undefined) {
      return [];
    }
    return Object.keys(props)
      .filter((key: string) => variants[key])
      .map((key: string) => props[key]);
  }
}
