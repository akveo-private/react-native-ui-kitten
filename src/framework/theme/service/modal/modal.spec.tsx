import React from 'react';
import {
  fireEvent,
  render,
} from 'react-native-testing-library';
import {
  TouchableOpacity,
  View,
  Text,
  Button,
} from 'react-native';
import { ModalService } from './modal.service';
import { ModalPanel } from '../../../theme/component';

jest.useFakeTimers();

describe('@modal-service: service checks', () => {

  const ShowSingleModalTestId: string = '@modal/single';
  const ShowMultipleModalTestId: string = '@modal/multiple';
  const ShowModalWithBackDropAllowedId: string = '@modal/backdrop';
  const ModalTextTestId: string = '@modal/modal-text';
  const TestModal2Text: string = 'Test Modal 2';

  class TestApplication extends React.Component<any, any> {

    showModal: () => void = () => {
      const component =
        <TestModal
          text={'Test Modal 1'}
          onCloseModal={this.props.onCloseModal}
          textTestId={ModalTextTestId}
        />;
      ModalService.showDialog(component);
    };

    showMultipleModals: () => void = () => {
      const component1 =
        <TestModal
          text={TestModal2Text}
          onCloseModal={this.props.onCloseModal}
          textTestId={`${ModalTextTestId}-1`}
        />;
      const component2 =
        <TestModal
          text={'Test Modal 2'}
          onCloseModal={this.props.onCloseModal}
          textTestId={`${ModalTextTestId}-2`}
        />;
      ModalService.showDialog(component1);
      ModalService.showDialog(component2);
    };

    showBackDropAllowedModal: () => void = () => {
      const component =
        <TestModal
          text={'Test Modal Back Drop'}
          onCloseModal={this.props.onCloseModal}
          textTestId={ModalTextTestId}
        />;
      ModalService.showDialog(component, true);
    };

    render() {
      return (
        <ModalPanel>
          <View>
            <TouchableOpacity testID={ShowSingleModalTestId} onPress={this.showModal}>
              <Text>Show Single Modal</Text>
            </TouchableOpacity>
            <TouchableOpacity testID={ShowMultipleModalTestId} onPress={this.showMultipleModals}>
              <Text>Show Single Modal</Text>
            </TouchableOpacity>
            <TouchableOpacity testID={ShowModalWithBackDropAllowedId} onPress={this.showBackDropAllowedModal}>
              <Text>Show Back-Drop Allowed Modal</Text>
            </TouchableOpacity>
          </View>
        </ModalPanel>
      );
    }
  }

  interface TestModalProps {
    text: string;
    textTestId: string;
    onCloseModal?: () => void;
  }

  class TestModal extends React.Component<TestModalProps> {

    render() {
      return (
        <View style={{ width: 300, height: 300, backgroundColor: 'red' }}>
          <Text testID={this.props.textTestId}>{this.props.text}</Text>
          <Button title={'Close Modal'} onPress={this.props.onCloseModal}/>
        </View>
      );
    }
  }

  it('* showDialog have been called', () => {
    const spy = jest.spyOn(ModalService, 'showDialog');
    const application = render(<TestApplication/>);
    fireEvent.press(application.getByTestId(ShowSingleModalTestId));

    expect(spy).toHaveBeenCalled();
  });

  it('* show multiple modals one by one', () => {
    const application = render(<TestApplication/>);
    fireEvent.press(application.getByTestId(ShowMultipleModalTestId));
    const expectedText: string = application
      .getByTestId(`${ModalTextTestId}-2`).props.children;

    expect(expectedText).toBe(TestModal2Text);
  });

  it('* unexpected branch cover', () => {
    const application = render(<TestApplication/>);
    ModalService.setComponent(null);
    expect(ModalService.component).toBe(null);
    fireEvent.press(application.getByTestId(ShowSingleModalTestId));
    expect(application).toMatchSnapshot();
  });

});
