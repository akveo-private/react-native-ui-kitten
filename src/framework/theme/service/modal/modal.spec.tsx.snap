// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@modal-service: service checks * show multiple modals one by one 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="@modal/single"
    >
      <Text>
        Show Single Modal
      </Text>
    </View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="@modal/multiple"
    >
      <Text>
        Show Single Modal
      </Text>
    </View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="@modal/backdrop"
    >
      <Text>
        Show Back-Drop Allowed Modal
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "position": "absolute",
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "red",
          "height": 300,
          "width": 300,
        }
      }
    >
      <Text
        testID="@modal/modal-text-1"
      >
        Test Modal 2
      </Text>
      <View
        accessibilityRole="button"
        accessibilityStates={Array []}
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Close Modal
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "position": "absolute",
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "red",
          "height": 300,
          "width": 300,
        }
      }
    >
      <Text
        testID="@modal/modal-text-2"
      >
        Test Modal 2
      </Text>
      <View
        accessibilityRole="button"
        accessibilityStates={Array []}
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Close Modal
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`@modal-service: service checks * showDialog have been called 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="@modal/single"
    >
      <Text>
        Show Single Modal
      </Text>
    </View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="@modal/multiple"
    >
      <Text>
        Show Single Modal
      </Text>
    </View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="@modal/backdrop"
    >
      <Text>
        Show Back-Drop Allowed Modal
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "position": "absolute",
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "red",
          "height": 300,
          "width": 300,
        }
      }
    >
      <Text
        testID="@modal/modal-text"
      >
        Test Modal 1
      </Text>
      <View
        accessibilityRole="button"
        accessibilityStates={Array []}
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Close Modal
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
